src/test/lib/scheduler/scheduler_test.cpp

```diff
@@ -126,113 +132,137 @@ TEST_F(SchedulerTest, BasicTest) {
-
-  // We expect NUM_GROUPS chains of tasks to be created. As tasks are added to the chains by calling
-  // AbstractTask::set_predecessor_of, the first task in the input vector ends up being the last task being called. This
-  // results in [40 30 20 10 0 41 31 21 11 1 ...]
-  const auto num_groups = NodeQueueScheduler::NUM_GROUPS;
-  EXPECT_EQ(TASK_COUNT % num_groups, 0);
-  auto expected_output = std::vector<size_t>{};
-  for (auto group = 0; group < num_groups; ++group) {
-    for (auto task_id = 0; task_id < TASK_COUNT / num_groups; ++task_id) {
-      expected_output.emplace_back(tasks.size() - (task_id + 1) * num_groups + group);
-    }
-  }
-
-  EXPECT_EQ(output, expected_output);
 }
 
 TEST_F(SchedulerTest, MultipleDependenciesWithScheduler) {
   Hyrise::get().topology.use_fake_numa_topology(8, 4);
   Hyrise::get().set_scheduler(std::make_shared<NodeQueueScheduler>());
 
-  std::atomic_uint32_t counter{0u};
+  auto counter = std::atomic_uint32_t{0};
 
   stress_multiple_dependencies(counter);
 
-  Hyrise::get().scheduler()->finish();
-
-  ASSERT_EQ(counter, 4u);
+  ASSERT_EQ(counter, 4);
 }
 
 TEST_F(SchedulerTest, DiamondDependenciesWithScheduler) {
   Hyrise::get().topology.use_fake_numa_topology(8, 4);
   Hyrise::get().set_scheduler(std::make_shared<NodeQueueScheduler>());
 
-  std::atomic_uint32_t counter{0};
+  auto counter = std::atomic_uint32_t{0};
 
   stress_diamond_dependencies(counter);
 
-  Hyrise::get().scheduler()->finish();
-
-  ASSERT_EQ(counter, 7u);
+  ASSERT_EQ(counter, 7);
 }
 
 TEST_F(SchedulerTest, LinearDependenciesWithoutScheduler) {
-  std::atomic_uint32_t counter{0u};
+  auto counter = std::atomic_uint32_t{0};
   stress_linear_dependencies(counter);
-  ASSERT_EQ(counter, 3u);
+  ASSERT_EQ(counter, 3);
 }
 
 TEST_F(SchedulerTest, MultipleDependenciesWithoutScheduler) {
-  std::atomic_uint32_t counter{0u};
+  auto counter = std::atomic_uint32_t{0};
   stress_multiple_dependencies(counter);
-  ASSERT_EQ(counter, 4u);
+  ASSERT_EQ(counter, 4);
 }
 
 TEST_F(SchedulerTest, DiamondDependenciesWithoutScheduler) {
-  std::atomic_uint32_t counter{0};
+  auto counter = std::atomic_uint32_t{0};
   stress_diamond_dependencies(counter);
-  ASSERT_EQ(counter, 7u);
+  ASSERT_EQ(counter, 7);
+}
+
+TEST_F(SchedulerTest, SuccessorExpired) {
+  if constexpr (!HYRISE_DEBUG) {
+    GTEST_SKIP();
+  }
+
+  const auto task1 = std::make_shared<JobTask>([&]() {});
+  const auto task2 = std::make_shared<JobTask>([&]() {});
+  task1->set_as_predecessor_of(task2);
+
+  {
+    const auto task3 = std::make_shared<JobTask>([&]() {});
+    task2->set_as_predecessor_of(task3);
+  }
+
+  task1->schedule();
+  // When task2 finishes, it will not be able to obtain its successor task3 as it went out of scope.
+  ASSERT_THROW(task2->schedule(), std::logic_error);
+
+  std::this_thread::sleep_for(std::chrono::milliseconds(40));
+  Hyrise::get().scheduler()->finish();
+}
+
+TEST_F(SchedulerTest, PassAllDependencies) {
```

Comment:
Can you add a comment what's happening here? Took me some seconds to see what's going on.

Reply:
```

TEST_F(SchedulerTest, NotAllDependenciesPassedToScheduler) {
  if constexpr (!HYRISE_DEBUG) {
    GTEST_SKIP();
  }

  const auto task1 = std::make_shared<JobTask>([&]() {});
  const auto task2 = std::make_shared<JobTask>([&]() {});
  const auto task3 = std::make_shared<JobTask>([&]() {});

  task1->set_as_predecessor_of(task2);
  task2->set_as_predecessor_of(task3);

  const auto tasks = std::vector<std::shared_ptr<AbstractTask>>{task1, task2};

  // The scheduler should complain that not all dependencies (task3 is a successor of task2) are passed.
  ASSERT_THROW(Hyrise::get().scheduler()->schedule_and_wait_for_tasks(tasks), std::logic_error);
}

```